### PR TITLE
[Nested Tensor] fix from_padded bug

### LINKED ISSUE #84082: Investigate from_padded implementations correctness
## Summary
A gradient test involving these ops created here: #84078 is succeeding for float64 on cuda but failing for float16 and float32 on cuda. The input sizes are small enough were floating point accumulation should not be effecting this tests.

## Env Details:
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.3
Libc version: glibc-2.27

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1069-aws-x86_64-with-glibc2.27
Is CUDA available: N/A
CUDA runtime version: 11.6.112
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torchdynamo==1.13.0.dev0
[conda] mkl                       2022.0.1           h06a4308_117
[conda] mkl-include               2022.0.1           h06a4308_117
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torchdynamo               1.13.0.dev0               dev_0    <develop>

cc @ezyang @gchanan @zou3519 @cpuhrsch @jbschlosser @bhosmer @mikaylagawarecki

### PR DESCRIPTION
**Pull request by drisspg**
Fixes #84082

Explained in the issue that the problem was arising from grad being not contiguous and the fast kernel not handiling this case gracefully.  The other thing I can do is add a contiguous call to https://github.com/pytorch/pytorch/blob/d144594512e10ab2a9625347816c2dee1fb55667/aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp#L45

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
If `_nested_from_padded` is interpreting the data as being contiguous, it better either assert that the input is so or make it contiguous itself. I don't think this is the right fix.

**Comment by drisspg** [comment]
> If `_nested_from_padded` is interpreting the data as being contiguous, it better either assert that the input is so or make it contiguous itself. I don't think this is the right fix.

Yup I have that fix locally,.Scott suggested offline, either call contiguous or instead if padded is not contiguous call the generic kernel. I am testing locally to see which is has better performance

**Comment by drisspg** [comment]
```
Using generic kernel if not contiguous:
  Median: 188.44 us
  IQR:    0.79 us (188.01 to 188.80)
  53 measurements, 1000 runs per measurement, 48 threads

Adding contiguous calls to custom kernel
  Median: 121.37 us
  IQR:    0.04 us (121.35 to 121.39)
  83 measurements, 1000 runs per measurement, 48 threads
```

script:
``` Python
import warnings
import torch
import random
import numpy as np
import torch.utils.benchmark as benchmark

np.random.seed(0)
random.seed(0)
torch.manual_seed(0)

warnings.filterwarnings("ignore")

if __name__ == "__main__":
    device='cuda'
    dtype=torch.float16

    a = torch.randn(512, 1024, device=device, dtype=dtype, requires_grad=False)
    b = torch.randn(1024, 1024, device=device, dtype=dtype, requires_grad=True)
    x = torch.nested_tensor([a, b])
    nt_size = x._nested_tensor_size()

    x = x.to_padded_tensor(0)
    x = x.transpose(-1,-2)
    # x = x.contiguous()

    torch._nested_from_padded(x, nt_size)
    t0 = benchmark.Timer(
    stmt='torch._nested_from_padded(x, nt_size)',
    globals={'x': x, 'nt_size': nt_size},
    label=f'Testing generic kernel with contiguous calls',
    num_threads=torch.get_num_threads())
    m0 = t0.blocked_autorange(min_run_time=10)
    print(m0)
 ```

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2702,7 +2702,7 @@
   cpu_nested_shape_example: non_differentiable
 
 - name: to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
-  self: at::_nested_from_padded(grad, self._nested_tensor_size())
+  self: at::_nested_from_padded(grad.contiguous(), self._nested_tensor_size())
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:2705]: This can be removed now?

Comment (drisspg) [reply tools/autograd/derivatives.yaml:2705]: Yes this can be removed now. Do you think having it is over cautious? Or do we have to trust the contract that nested_from_padded provides hence we don't need it, and adding it is confusing.

Comment (albanD) [reply tools/autograd/derivatives.yaml:2705]: I think it is confusing because it gives the false impression that this function requires contiguous inputs.

Comment (drisspg) [reply tools/autograd/derivatives.yaml]: Yeah that makes sense, just pushed the change

**Review by albanD — APPROVED** [review]
SGTM!
